### ISSUE TITLE
Unsupported parent methods in MultiChannelGrid3D

### ISSUE THREAD

**Issue by tmgeimer** [akmaier/CONRAD#68]
MultiChannelGrid3D inherits from Grid3D but instead of using the internal buffer of its parent class, it used a Grid4D member to implement vector valued voxels.
```java
public class MultiChannelGrid3D extends Grid3D {

	String [] channelNames;
	Grid4D multichannelData;
	
	public MultiChannelGrid3D(int width, int height, int depth, int channels) {
		// This will cause some parent methods to fail ( e.g. getSize() );
		// TODO Override necessary parent methods.
		super(0,0,0);
		multichannelData = new Grid4D(width, height, depth, channels);
		buffer = multichannelData.getSubGrid(0).getBuffer();
		//this(width, height, depth, true);
	}
```
However, implementing it this way has two major downsides:

1. The internal buffer of Grid3D is initialized in super(0,0,0). However, being a Grid3D, MultiChannelGrid3D provides parent methods, that do not work on multi channel data, as they access the internal buffer which is never filled nor used. Prior to my changes, getSize() would always return [0,0,0] and not as one would expect the size of multiChannelData. Many more parent function are not working properly atm.

2. Realizing a vector valued image this way may cause "jumping" within the memory, as the channels represent the most outer dimension (i.e. each channel is represented in a separate Grid3D). While this is suitable for RGB images to separate the channels, it may not be a good choice for realizing grids, that feature vector valued voxels (e.g. deformation fields). Here, having the vector entries as the most inner dimension could be preferable for easier memory access.

MultiChannelGrid2D poses the same problems (being a Grid2D with a Grid3D member).
Possible solutions: Either override all Grid3D methods in MultiChannelGrid3D to properly handle multi-channel data, or implement a generic MultiChannelGridND (possibly inheriting from NumericGrid) class for MultiChannelGrid3D / MultiChannelGrid2D to inherit from.

**Comment by bergerma** [comment]
A similar topic has been discusses about one year ago. As a result I decided to come up with the "generic" grid class package, that allows ND grids with data types defined by java generics. The implementing user has to deal with the actual storage or mapping to linear memory. Please look at the ComplexGrid classes there to get an impression about the structure. For the complex classes we decided to map them down to linear memory to ease GPU functionality but also to make them compatible with the existing FFT library, yet, for other types you could think of different memory layouts, e.g. ArrayLists or other containers.

Anyway, there is a good reason why the "MultiChannelGridXD" classes derive from apparently lower dimensional grid classes. The whole parallel pipelining scheme in CONRAD is based on the Grid2D class, as the vast majority of parallel operations is done image-wise, e.g. for each projection image. Deriving the MultiChannelGrid2D class from Grid2D allowed its usage in the parallel processing pipeline. In case of the MultiChannelGrid3D class this structure was kept the same, however, the parallel processing is not available for Grid3Ds anyway so we could indeed argue if it makes sense. A change of the MultiChannelGrids to ND grids with arbitrary types would require a major refactoring of the whole processing structure in CONRAD and would require profound planning.

**Comment by akmaier** [comment]
Indeed. If we want to change this, it will come with several side effects. The main reason for using Grid2D for the pipeline interface is a legacy from using ImageJ's ImageProcessor. In order to replace it, we would have to create a series of Grid interfaces that specify which Grid type the tool is able to process. So in conclusion, I think refactoring this is possible, but it will affect most of the reconstruction classes. We probably need a Hackathon to refactor and test that we did not break anything...
